internal/topic/topicreadercommon/decoders.go

```diff
@@ -5,43 +5,122 @@ import (
 	"errors"
 	"fmt"
 	"io"
+	"sync"
 
 	"github.com/ydb-platform/ydb-go-sdk/v3/internal/grpcwrapper/rawtopic/rawtopiccommon"
 	"github.com/ydb-platform/ydb-go-sdk/v3/internal/xerrors"
 )
 
+type ReadResetter interface {
+	io.Reader
+	Reset(r io.Reader) error
+}
+
+type decoderPool struct {
+	pool sync.Pool
+}
+
+func (p *decoderPool) Get() ReadResetter {
+	dec, _ := p.pool.Get().(ReadResetter)
+
+	return dec
+}
+
+func (p *decoderPool) Put(dec ReadResetter) {
+	p.pool.Put(dec)
+}
+
+func newDecoderPool() *decoderPool {
+	return &decoderPool{
+		pool: sync.Pool{},
+	}
+}
+
 type DecoderMap struct {
-	m map[rawtopiccommon.Codec]PublicCreateDecoderFunc
+	m  map[rawtopiccommon.Codec]func(io.Reader) (ReadResetter, error)
+	dp map[rawtopiccommon.Codec]*decoderPool
 }
 
 func NewDecoderMap() DecoderMap {
-	return DecoderMap{
-		m: map[rawtopiccommon.Codec]PublicCreateDecoderFunc{
-			rawtopiccommon.CodecRaw: func(input io.Reader) (io.Reader, error) {
-				return input, nil
-			},
-			rawtopiccommon.CodecGzip: func(input io.Reader) (io.Reader, error) {
-				return gzip.NewReader(input)
-			},
-		},
+	dm := DecoderMap{
+		m:  make(map[rawtopiccommon.Codec]func(io.Reader) (ReadResetter, error)),
+		dp: make(map[rawtopiccommon.Codec]*decoderPool),
 	}
+
+	dm.AddDecoder(rawtopiccommon.CodecRaw, func(input io.Reader) (ReadResetter, error) {
+		return &nopResetter{Reader: input}, nil
+	})
+
+	dm.AddDecoder(rawtopiccommon.CodecGzip, func(input io.Reader) (ReadResetter, error) {
+		gz, err := gzip.NewReader(input)
+		if err != nil {
+			return nil, err
+		}
+
+		return gz, nil
+	})
+
+	return dm
 }
 
-func (m *DecoderMap) AddDecoder(codec rawtopiccommon.Codec, createFunc PublicCreateDecoderFunc) {
+func (m *DecoderMap) AddDecoder(codec rawtopiccommon.Codec, createFunc func(io.Reader) (ReadResetter, error)) {
 	m.m[codec] = createFunc
+	m.dp[codec] = newDecoderPool()
```

Comment:
What about support custom codec?

We must allow to accept custom external codec and save current interface.
For test you can create custom codec with read bytes in reverse order.

We should work with items, created with createFunc. Pool must be usable for custom codec too (for example -check implementation of ReadResetter before store item to pool).